simulators/ethereum/pyspec/runner.go

```diff
@@ -99,123 +92,110 @@ func (tc *testcase) run(t *hivesim.T) {
-		}
-		// check final storage
-		if len(genesisAccount.Storage) > 0 {
-			// extract fixture storage keys
-			keys := make([]common.Hash, 0, len(genesisAccount.Storage))
-			for key := range genesisAccount.Storage {
-				keys = append(keys, key)
-			}
-			// get storage values for account with keys: keys
-			gotStorage, errS := engineClient.StorageAtKeys(ctx, account, keys, nil)
-			if errS != nil {
-				tc.failedErr = errS
-				t.Errorf("unable to get storage values from account: %v, in test %s: %v", account, tc.name, errS)
+	if err := tc.ValidatePost(ctx, engineClient); err != nil {
+		tc.Fatalf("unable to verify post allocation in test %s: %v", tc.Name, err)
+	}
+
+	if tc.SyncPayload != nil {
+		// First send a new payload to the already running client
+		if syncing, err := tc.SyncPayload.ExecuteValidate(
+			ctx,
+			engineClient,
+		); err != nil {
+			tc.Fatalf("unable to send sync payload: %v", err)
+		} else if syncing {
+			tc.Fatalf("sync payload failed (not synced)")
+		}
+		// Send a forkchoice update to the already running client to head to the sync payload
+		if syncing, err := tc.SyncPayload.ForkchoiceValidate(ctx, engineClient, tc.EngineFcuVersion); err != nil {
+			tc.Fatalf("unable to update head of chain: %v", err)
+		} else if syncing {
+			tc.Fatalf("forkchoice update failed (not synced)")
+		}
+
+		// Spawn a second client connected to the already running client,
+		// send the forkchoice updated with the head hash and wait for sync.
+		// Then verify the post allocation.
+		// Add a timeout too.
+		secondEngineClient, err := engineStarter.StartClient(t, ctx, tc.Genesis(), env, nil, engineClient)
+		if err != nil {
+			tc.Fatalf("can't start client with Engine API: %v", err)
+		}
+
+		if _, err := tc.SyncPayload.ExecuteValidate(
+			ctx,
+			secondEngineClient,
+		); err != nil {
+			tc.Fatalf("unable to send sync payload: %v", err)
+		} // Don't check syncing here because some clients do sync immediately
+
+		timeoutCtx, cancel := context.WithTimeout(ctx, SyncTimeout)
+		defer cancel()
+		for {
+			if syncing, err := tc.SyncPayload.ForkchoiceValidate(ctx, secondEngineClient, tc.EngineFcuVersion); err != nil {
+				tc.Fatalf("unable to update head of chain: %v", err)
+			} else if !syncing {
+				break
 			}
-			// check values in storage match with fixture
-			for _, key := range keys {
-				if genesisAccount.Storage[key] != *gotStorage[key] {
-					tc.failedErr = errors.New("storage received doesn't match expected from fixture")
-					t.Errorf(`storage received from account %v doesn't match expected from fixture in test %s: from storage address: %v
-						received from block:  %v
-						expected in fixture:  %v`, account, tc.name, key, gotStorage[key], genesisAccount.Storage[key])
-				}
+			select {
+			case <-timeoutCtx.Done():
+				tc.Fatalf("timeout waiting for sync of secondary client")
+			default:
 			}
+			time.Sleep(time.Second)
 		}
 	}
+
 	end := time.Now()
 
```

Comment:
I don't think it would hurt if we used `ValidatePost()` here for the second client as well

Reply:
The problem was that, technically it's not the same post, this is due to the extra syncing block.
So I was a little bit afraid to activate the post check here without giving it enough thought.

We could revisit it once we start marking some more tests as sync-verified.